packages/tracing-internal/src/browser/instrument.ts

```diff
@@ -0,0 +1,229 @@
+import { getFunctionName, logger } from '@sentry/utils';
+
+import { onCLS } from './web-vitals/getCLS';
+import { onFID } from './web-vitals/getFID';
+import { onLCP } from './web-vitals/getLCP';
+import { observe } from './web-vitals/lib/observe';
+
+type InstrumentHandlerTypePerformanceObserver = 'longtask' | 'event' | 'navigation' | 'paint' | 'resource';
+
+// We provide this here manually instead of relying on a global, as this is not available in non-browser environements
+// And we do not want to expose such types
+interface PerformanceEntry {
+  readonly duration: number;
+  readonly entryType: string;
+  readonly name: string;
+  readonly startTime: number;
+  toJSON(): Record<string, unknown>;
+}
+
+interface Metric {
+  /**
+   * The name of the metric (in acronym form).
+   */
+  name: 'CLS' | 'FCP' | 'FID' | 'INP' | 'LCP' | 'TTFB';
+
+  /**
+   * The current value of the metric.
+   */
+  value: number;
+
+  /**
+   * The rating as to whether the metric value is within the "good",
+   * "needs improvement", or "poor" thresholds of the metric.
+   */
+  rating: 'good' | 'needs-improvement' | 'poor';
+
+  /**
+   * The delta between the current value and the last-reported value.
+   * On the first report, `delta` and `value` will always be the same.
+   */
+  delta: number;
+
+  /**
+   * A unique ID representing this particular metric instance. This ID can
+   * be used by an analytics tool to dedupe multiple values sent for the same
+   * metric instance, or to group multiple deltas together and calculate a
+   * total. It can also be used to differentiate multiple different metric
+   * instances sent from the same page, which can happen if the page is
+   * restored from the back/forward cache (in that case new metrics object
+   * get created).
+   */
+  id: string;
+
+  /**
+   * Any performance entries relevant to the metric value calculation.
+   * The array may also be empty if the metric value was not based on any
+   * entries (e.g. a CLS value of 0 given no layout shifts).
+   */
+  entries: PerformanceEntry[];
+
+  /**
+   * The type of navigation
+   *
+   * Navigation Timing API (or `undefined` if the browser doesn't
+   * support that API). For pages that are restored from the bfcache, this
+   * value will be 'back-forward-cache'.
+   */
+  navigationType: 'navigate' | 'reload' | 'back-forward' | 'back-forward-cache' | 'prerender';
+}
+
+export type InstrumentHandlerType = 'cls' | 'lcp' | 'fid' | InstrumentHandlerTypePerformanceObserver;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type InstrumentHandlerCallback = (data: any) => void;
+
+type CleanupHandlerCallback = () => void;
+
+const handlers: { [key in InstrumentHandlerType]?: InstrumentHandlerCallback[] } = {};
+const instrumented: { [key in InstrumentHandlerType]?: boolean } = {};
+
+/** Instruments given API */
+function instrument(type: InstrumentHandlerType): void {
```

Comment:
I'd like to see instrument broken down into individual exported functions instead of using the switch statement (reduces bundle size), but I'm fine with keeping it for now.

Reply:
Hmm, I wonder if it would really reduce bundle size, as we'd still need to do `instrumented['cls']` etc. in each of the functions? But I can give it a try!

Reply:
OK, I split this up, which is probably nicer anyhow. I was able to keep this in as reusable pieces as possible, making this as small as can be (and I think it shaved off a few bytes even).